pkg/streamingpromql/operators/aggregations/limitklimitratio/operator.go

```diff
@@ -0,0 +1,509 @@
+}
+
+func (s *limitStepCounter) count(step int) int {
+	return s.seriesIncludedAtStep[step]
+}
+
+func newStepCounter(size int, memoryConsumptionTracker *limiter.MemoryConsumptionTracker) (*limitStepCounter, error) {
+	intSlice, err := types.IntSlicePool.Get(size, memoryConsumptionTracker)
+	if err != nil {
+		return nil, err
+	}
+	intSlice = intSlice[:size]
+	return &limitStepCounter{
+		memoryConsumptionTracker: memoryConsumptionTracker,
+		seriesIncludedAtStep:     intSlice,
+	}, nil
+}
+
+type groupLimiter interface {
+	close()
+
+	incSeriesRefCount()
+
+	// initSeries informs the groupLimiter that this series will be included in it's grouping
+	initSeries(seriesIndex int, series types.SeriesMetadata)
+
+	// releaseSeries informs the groupLimiter that this series has finished being accumulated and any internal resources related to this series can be released
+	releaseSeries(seriesIndex int)
+
+	// startAccumulatingSeries is invoked each time we start accumulating data for a series.
+	// if false is returned then this series should not be accumulated - it can be discarded
+	startAccumulatingSeries() bool
+
+	// onStepAccumulated is invoked for each step in accumulating the data for a series
+	onStepAccumulated(step int, hasValue bool)
+
+	// endAccumulatingSeries is invoked once we have finished accumulating the series data
+	endAccumulatingSeries()
+
+	// discardSampleAtStep returns true if the sample at this step should be ignored/discarded
+	discardSampleAtStep(seriesIndex int, step int) bool
+}
+
+type groupLimiterRatio struct {
+	seriesRefCount int // Track how many series are related to this group.
+	stepArg        *stepArgumentRatio
+	seriesHashMap  map[int]float64
+}
+
+func (q *groupLimiterRatio) incSeriesRefCount() {
+	q.seriesRefCount++
+}
+
+func (q *groupLimiterRatio) startAccumulatingSeries() bool   { return true }
+func (q *groupLimiterRatio) onStepAccumulated(_ int, _ bool) {}
+func (q *groupLimiterRatio) endAccumulatingSeries()          {}
+func (q *groupLimiterRatio) close()                          {}
+
+func (q *groupLimiterRatio) initSeries(seriesIndex int, series types.SeriesMetadata) {
+
+	if q.seriesHashMap == nil {
+		q.seriesHashMap = make(map[int]float64, q.seriesRefCount)
+	}
+
+	// This hash matches the prometheus implementation of promql.HashRatioSampler in engine.go
+	// TODO update promql.HashRatioSampler to expose this hash function and call directly into this rather then calculate the hash here
+	// The reason for extracting this out is that currently the promql.HashRatioSampler does not expose this hash function directly,
+	// which means we need to call both the hash and ratio comparison for every step. This code path allows for the hash to be
+	// calculated once per series and referenced in each step comparison.
+	const (
+		float64MaxUint64 = float64(math.MaxUint64)
+	)
+	sample := &promql.Sample{Metric: series.Labels}
+	q.seriesHashMap[seriesIndex] = float64(sample.Metric.Hash()) / float64MaxUint64
```

Comment:
Why do we need to construct a `promql.Sample` here? Can't we just do this?
```suggestion
	q.seriesHashMap[seriesIndex] = float64(series.Labels.Hash()) / float64MaxUint64
```